subprojects/core/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheEntryExpiration.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.caching.local.internal;
+
+import org.gradle.api.internal.cache.CacheConfigurationsInternal;
+import org.gradle.caching.BuildCacheServiceFactory;
+import org.gradle.caching.local.DirectoryBuildCache;
+import org.gradle.internal.time.TimeFormatting;
+import org.gradle.internal.time.TimestampSuppliers;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class DirectoryBuildCacheEntryExpiration {
+    private final Supplier<Long> expiryTime;
+    private final String description;
+
+    public DirectoryBuildCacheEntryExpiration(DirectoryBuildCache directoryBuildCache, CacheConfigurationsInternal cacheConfigurations) {
+        @SuppressWarnings("deprecation")
+        int deprecatedRemoveUnusedEntriesAfter = directoryBuildCache.getRemoveUnusedEntriesAfterDays();
+
+        // Use the deprecated retention period if configured on `DirectoryBuildCache`, or use the core 'buildCache' cleanup configuration if not.
+        // If the deprecated property remains at the default, we can safely use the central value (which has the same default).
+        if (deprecatedRemoveUnusedEntriesAfter == CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_BUILD_CACHE_ENTRIES) {
+            this.expiryTime = cacheConfigurations.getBuildCache().getRemoveUnusedEntriesOlderThanAsSupplier();
+            long expiryDuration = System.currentTimeMillis() - this.expiryTime.get();
```

Comment:
🤔 I find it somewhat strange that the configuration value seems to depend on the current timestamp. I suppose it does depend a lot on how you configure the new parameter.

Reply:
Yes, the expiry can still be configured as `removeUnusedEntriesAfterDays(5)`, but we wanted to be flexible and allow an arbitrary timestamp to be configured.

So we're effectively converting the "number of days" to a timestamp, and then reverse engineering this to present a human-consumable value to Build Scans.

Reply:
Is the configuration re-run on a configuration cache hit? Or is the value stored as is in the configuration cache? If it isn't re-run, then `removeUnusedEntriesAfterDays(5)` might mean `removeUnusedEntriesAfterDays(10)` if you had configuration cache hits for 5 consecutive days.

> So we're effectively converting the "number of days" to a timestamp, and then reverse engineering this to present a human-consumable value to Build Scans.

It would be good if the configuration exposes whether the timestamp is supposed to be relative or absolute. Though I am not sure how hard that would be to implement. Build Scans do not rely on name of the key, so we could also change the key if that makes it better. IMO it would be good if the configuration by build scans doesn't change if the user doesn't change the configuration.

Reply:
> Is the configuration re-run on a configuration cache hit? Or is the value stored as is in the configuration cache?

We do the translation of user configuration to timestamp when constructing the BuildCache service, so I'm pretty confident the timestamp itself won't be persisted in the configuration cache. But I will test this.

> It would be good if the configuration exposes whether the timestamp is supposed to be relative or absolute.

I don't think I understand. The timestamp is an instant expressed in milliseconds since epoch. It's always going to be absolute.

The idea was that we don't tie our internal infrastructure to the configuration API: we should be able to provide other configuration mechanisms (eg `removeEntriesUnusedSinceStartOfBuild()`) and have these converted to a timestamp for the infrastructure to deal with.

> Build Scans do not rely on name of the key, so we could also change the key if that makes it better. IMO it would be good if the configuration by build scans doesn't change if the user doesn't change the configuration.

Can you clarify how you'd like to see things expressed in Build Scans, assuming that there may be different ways of configuring the expiry in the future? We could simply render the timestamp directly (`removeEntriesNotUsedSince`) but I think this won't be better for users.

Reply:
> I don't think I understand. The timestamp is an instant expressed in milliseconds since epoch. It's always going to be absolute.
>
> The idea was that we don't tie our internal infrastructure to the configuration API: we should be able to provide other configuration mechanisms (eg removeEntriesUnusedSinceStartOfBuild()) and have these converted to a timestamp for the infrastructure to deal with.

Yes, the timestamp is absolute, though the configuration API allows you to provide relative timestamps. And you expect for every build that the relative timestamp (aka older than 5 days) to be re-evaluated to an absolute timestamp. I'd somewhat expect whatever we send to scans to reflect the semantic meaning of the configuration API, not the internal value. And if we add something new to the configuration we could reflect that in scans as well.

Though like I said, I am fine to use what we do in this PR.

Reply:
Actually, the [configuration API only allows you to specify number of days](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home:configure_cache_cleanup), just like the deprecated mechanism on `DirectoryBuildCache`.
This value is [immediately converted to a timestamp](https://github.com/gradle/gradle/blob/31a349ecf6827ba52baa59c9f1340de493883822/subprojects/core/src/main/java/org/gradle/api/internal/cache/DefaultCacheConfigurations.java#L245), which makes me wonder about the impact of configuration-cache on these settings.

I'll investigate.